### PR TITLE
Initial work on github publish

### PR DESCRIPTION
**Pull request by jdetle**
## Todo
- [x] Figure out how to ensure things aren't breaking with the call to authenticate
- [x] Inserted some TODO comments @captainsafia or @rgbkrk if you want to go ahead and answer some of the questions I'd appreciate it. 
## Work Done

---
- Added Docs
- Added a call to github.authenticate() to get gist to no longer be anonymous
## Work left for another PR

---

I'd like to get username and password to be acceptable, it would require some sort of frontend tool for entry but it is easy on the backend.

### CODE REVIEW

File: src/notebook/epics/github-publish.js [hunk 1]
```diff
@@ -17,29 +17,47 @@ const Observable = Rx.Observable;
...
     } else {
       observer.next(new Github());
     }
     observer.complete();
   });
 
+/**
+ * TODO I have no idea what this does yet
+ * @param {}
+ */
```
Comment (rgbkrk) [inline src/notebook/epics/github-publish.js:49]: setGithub is an action creator (redux parlance). It returns a plain object that gets merged into the state tree for everyone.

Here it takes the GitHub auth object and creates proper action format for setting our GitHub context.

File: src/notebook/epics/github-publish.js [hunk 2]
```diff
@@ -49,6 +67,14 @@ export const initialGitHubAuthEpic = () => {
     .map(setGithub);
```
Comment (rgbkrk) [inline src/notebook/epics/github-publish.js:69]: The observable that epics return must emit redux actions like that of SET_GITHUB

File: src/notebook/epics/github-publish.js [hunk 3]
```diff
@@ -65,25 +91,45 @@ function notifyUser(filename, gistURL, gistID, notificationSystem) {
...
       observer.complete();
       return;
     }
-
     const gistID = response.id;
     const gistURL = response.html_url;
 
     notifyUser(filename, gistURL, gistID, notificationSystem);
-    if (hotOffThePresses) {
+    if (firstTimePublish) {
```
Comment (rgbkrk) [inline src/notebook/epics/github-publish.js:118]: Thanks. :)

File: src/notebook/epics/github-publish.js [hunk 4]
```diff
@@ -104,7 +150,12 @@ export function publishNotebookObservable(github, notebook, filepath, notificati
       message: 'Your notebook is being uploaded as a GitHub gist',
       level: 'info',
     });
-
+    // Hacky fix to authenticate user, according to docs this is synchronous
+    // and may break things, TODO, figure out how to make this not possibly
+    // break things
+    if (process.env.GITHUB_TOKEN) {
+      github.authenticate({ type: 'oauth', token: process.env.GITHUB_TOKEN });
```
Comment (rgbkrk) [inline src/notebook/epics/github-publish.js]: It's actually synchronous because it doesn't do the auth call - just storing it for later.

**Comment by jdetle** [comment]
Done. Some future work includes:
1. Saving a token as an object and using that for authentication. 
2. A dialog box for authenticating over the web would be nice.
3. A button for a live overwriting for for showing off a notebook live would be cool too.

**Comment by rgbkrk** [comment]
Incredibly great cleanup and fix here. 👍

**Pull request merged**